hourly-tips/src/main/java/org/apache/flink/training/exercises/hourlytips/HourlyTipsExercise.java

```diff
@@ -76,9 +84,19 @@ public JobExecutionResult execute() throws Exception {
         DataStream<TaxiFare> fares = env.addSource(source);
 
         // replace this with your solution
-        if (true) {
-            throw new MissingSolutionException();
-        }
+        DataStream<TaxiFare> watermarkedFares = fares.assignTimestampsAndWatermarks(
+                WatermarkStrategy
+                        .<TaxiFare>forMonotonousTimestamps()
+                        .withTimestampAssigner((taxiFare, oldTimeStamp) -> taxiFare.startTime.toEpochMilli())
+        );
+
+        DataStream<Tuple3<Long, Long, Float>> driversWithSummedTips = watermarkedFares.keyBy(fare -> fare.driverId)
+                .window(TumblingEventTimeWindows.of(Time.minutes(60)))
+                .reduce(new TipSum(), new ProcessTipSumWindowFunction());
```

Comment:
why reduce? (not process)
Recommend to check HourlyTipsSolution

Reply:
I have provided a second solution using process and used parametrized tests to run both.

I think the reduce solution is more natural for computing the sum and also has the benefit of keeping only the currendt value of the accumulator in memory instead of the entire window contents (`Iterable<TaxiFare> fares`)